Being able to select a detached head commit in the grid

pmiossec opened the pull request:
When we are in a detached head state,
the `git branch` command output "(HEAD detached at c299581)"
and the sha1 should be extracted to be able to select the good revision
instead of getting the ugly popup "no revision found!"

Way to reproduce:
Checkout a specific revision and use the left panel
in selecting/unselecting this "branch"

pmiossec committed:
Being able to select a detached head commit in the grid

When we are in a detached head state,
the `git branch` command output "(HEAD detached at c299581)"
and the sha1 should be extracted to be able to select the good revision
instead of getting the ugly popup "no revision found!"

pmiossec commented:
PS: You're MOD is a really good job and I wish a lot of your job will be already merged in the upstream repository. Unfortunately, development around GitExtensions has slowed down a lot since a few months :(

EbenZhang commented:
Thanks. I will review and test it this week.

pmiossec commented:
No problem. Feel free to cherrypick instead of merging my PR if it suits you more....

EbenZhang commented:
Don't worry it's not merge-able. I will cherry pick if needed.

EbenZhang commented:
On my it's `(detached from c299581)`, which version of git are you using?

pmiossec commented:
From memory, I think I use 2.6.2 (but at least >2.5.0) but I think I could
update the regex to catch the 2 strings...

What is your version of git? You stayed to 1.9.5?

EbenZhang commented:
We should support old version of git.
We may also need to add the new one to `DetachedPrefixes` in `GitModule.cs`

pmiossec committed:
Manage new format to detect detached HEAD

which is now returned by the `git branch` command output:
"(HEAD detached at c299581)"

See https://github.com/git/git/commit/4b06318664638d306cad920fd86eb63b69739310

pmiossec commented:
Should be fixed...

Note: the first commit has been pushed upstream (PR)

pmiossec commented:
I didn't know that the format had changed recently :(

EbenZhang commented on GitCommands/Git/GitModule.cs:
```diff
@@ -2542,6 +2542,24 @@ public static bool IsDetachedHead(string branch)
+        public static bool TryParseDetachedHead(string text, out string sha1)
+        {
+            sha1 = null;
+            if (IsDetachedHead(text))
```
Mising a `!` here.

pmiossec replied:
Oh,  sheet! You're right! I forgot it during my last refactoring... I'll fix it.

EbenZhang commented:
Thanks for the changes, I have cherry picked and fixed the typo of missing a `!`
